src/storage/btree.c

```diff
@@ -16977,15 +16999,42 @@ btree_rv_roothdr_undo_update (THREAD_ENTRY * thread_p, LOG_RCV * recv)
   root_header = btree_get_root_header (thread_p, recv->pgptr);
   assert (root_header != NULL);
 
+  num_nulls = (unsigned int) root_header->num_nulls;
+  num_oids = (unsigned int) root_header->num_oids;
+  num_keys = (unsigned int) root_header->num_keys;
+
+  if (root_header->_64.over)
+    {
+      num_nulls |= (long long) root_header->num_nulls << 32;
+      num_oids |= (long long) root_header->num_oids << 32;
+      num_keys |= (long long) root_header->num_keys << 32;
+    }
+
   if (root_header != NULL)
     {
+      int over = 0;
+      long long delta;
+
       /* unpack the root statistics */
       datap = (char *) recv->data;
-      root_header->num_nulls += OR_GET_INT (datap);
-      datap += OR_INT_SIZE;
-      root_header->num_oids += OR_GET_INT (datap);
-      datap += OR_INT_SIZE;
-      root_header->num_keys += OR_GET_INT (datap);
+      OR_GET_BIGINT (datap, &delta);
+      num_nulls += delta;
+      datap += OR_BIGINT_SIZE;
+      OR_GET_BIGINT (datap, &delta);
+      num_oids += delta;
+      datap += OR_BIGINT_SIZE;
+      OR_GET_BIGINT (datap, &delta);
+      num_keys += delta;
+
+      root_header->num_nulls = num_nulls & 0xffffffff;
+      root_header->num_oids = num_oids & 0xffffffff;
+      root_header->num_keys = num_keys & 0xffffffff;
+
+      over += root_header->_64.num_nulls = num_nulls >> 32;
+      over += root_header->_64.num_oids = num_oids >> 32;
+      over += root_header->_64.num_keys = num_keys >> 32;
+
+      root_header->_64.over = (over > 0);
```

Comment:
This routine is used several times. So, how about making macro or inline functions?

Reply:
I agree with you but I keep the style for easy debugging.